metaflow/plugins/kfp/kfp.py

```diff
@@ -1,86 +1,275 @@
+    """
+    Create a map of steps to their corresponding command templates. These command templates help define the command
+    to be used to run that particular step with placeholders for the `run_id` and `datastore_root` (location of the datastore).
+
+    # Note:
+    # Level-order traversal is adopted to keep the task-ids in line with what happens during a local metaflow execution.
+    # It is not entirely necessary to keep this order of task-ids if we are able to point to the correct input-paths for
+    # each step. But, using this ordering does keep the organization of data more understandable and natural (i.e., `start`
+    # gets a task id of 1, next step gets a task id of 2 and so on with 'end' step having the highest task id.
+    """
+    def build_cmd_template(step_name, task_id, input_paths):
+        python_cmd = "python {downloaded_file_name} --datastore s3 --datastore-root {{ds_root}} " \
+                     "step {step_name} --run-id {{run_id}} --task-id {task_id} " \
+                     "--input-paths {input_paths}".format(downloaded_file_name=DEFAULT_DOWNLOADED_FLOW_FILENAME,
+                                                            step_name=step_name, task_id=task_id, input_paths=input_paths)
+        return python_cmd
+
+    steps_deque = deque(['start']) # deque to process the DAG in level order
+    cur_task_id = 0
+
+    # set of seen steps, i.e., added to the queue for processing
+    seen_steps = set(['start'])
+    # Mapping of steps to task ids
+    task_id_map = {}
+    # Mapping of steps to their command templates
+    command_template_map = {}
+
+    while len(steps_deque) > 0:
+        cur_step = steps_deque.popleft()
+        cur_task_id += 1
+        task_id_map[cur_step] = cur_task_id
+
+        # Generate the correct input_path for each step. Note: input path depends on a step's parents (i.e., in_funcs)
+        # Format of the input-paths for reference:
+        # non-join nodes: run-id/parent-step/parent-task-id,
+        # branch-join node: run-id/:p1/p1-task-id,p2/p2-task-id,...
+        # foreach node: TODO: foreach is not considered here
+        if cur_task_id == 1: # start step
+            cur_input_path = '{run_id}/_parameters/0' # this is the standard input path for the `start` step
+        else:
+            if graph.nodes[cur_step].type != 'join':
+                parent_step = graph.nodes[cur_step].in_funcs[0]
+                cur_input_path = '{run_id}/'+ parent_step + '/' + str(task_id_map[parent_step])
+            else:
+                cur_input_path = '{run_id}/:'
+                for parent in graph.nodes[cur_step].in_funcs:
+                    cur_input_path += parent + '/' + str(task_id_map[parent]) + ','
+                cur_input_path = cur_input_path.strip(',')
+
+        command_template_map[cur_step] = build_cmd_template(cur_step, cur_task_id, cur_input_path)
+
+        out_funcs = graph.nodes[cur_step].out_funcs
+        for step in out_funcs:
```

Comment:
Just to confirm looks like `in_funcs` is a `List` but `out_funcs` is a `Dict`?

Reply:
No, both are lists containing the corresponding names of the steps

Reply:
Cool okay, might be worthwhile renaming `step` to `step_idx` or something so it's a little clearer, I assumed it was a dictionary and you were iterating over the keys since you were using it to concatenate a string 😅